src/main/index.ts

```diff
@@ -523,24 +525,59 @@ const router = t.router({
     const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
 
     if (happAndUiBytes.uiBytes) {
-      await holochainManager.installWebHapp(happAndUiBytes, appId, networkSeed);
+      await holochainManager.installWebHappFromBytes(happAndUiBytes, appId, networkSeed);
     } else {
-      await holochainManager.installHeadlessHapp(happAndUiBytes.happBytes, appId, networkSeed);
+      await holochainManager.installHeadlessHappFromBytes(
+        happAndUiBytes.happBytes,
+        appId,
+        networkSeed,
+      );
     }
   }),
-  installHappFromBytes: t.procedure.input(InstallHappInputSchema).mutation(async (opts) => {
-    const { bytes, appId, networkSeed } = opts.input;
-
-    const happAndUiBytes = await rustUtils.decodeHappOrWebhapp(Array.from(bytes));
-
+  isHappAvailableAndValid: t.procedure.input(z.string()).mutation(async (opts) => {
     const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
-
-    if (happAndUiBytes.uiBytes) {
-      await holochainManager.installWebHapp(happAndUiBytes, appId, networkSeed);
-    } else {
-      await holochainManager.installHeadlessHapp(happAndUiBytes.happBytes, appId, networkSeed);
-    }
+    return holochainManager.isHappAvailableAndValid(opts.input);
+  }),
+  isUiAvailable: t.procedure.input(z.string()).mutation(async (opts) => {
+    const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
+    return holochainManager.isUiAvailable(opts.input);
   }),
+  storeUiBytes: t.procedure.input(BytesSchema).mutation(async (opts) => {
+    const { bytes } = opts.input;
+    const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
+    return holochainManager.storeUiIfNecessary(Array.from(bytes));
+  }),
+  storeHappBytes: t.procedure.input(BytesSchema).mutation(async (opts) => {
+    const { bytes } = opts.input;
+    const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
+    return holochainManager.storeHapp(Array.from(bytes));
+  }),
+  installWebhappFromHashes: t.procedure
+    .input(InstallWebhappFromHashesSchema) // TODO: need metadata input as well here like name and action hash of app and app version in app store
+    .mutation(async (opts) => {
+      const { happSha256, uiZipSha256, appId, networkSeed } = opts.input;
+      const holochainManager = getHolochainManager(HOLOCHAIN_DATA_ROOT!.name);
+      await holochainManager.installWebhappFromHashes(happSha256, uiZipSha256, appId, networkSeed);
+    }),
+  installWebhappFromBytes: t.procedure
+    .input(InstallHappOrWebhappFromBytesSchema)
+    .mutation(async (opts) => {
```

Comment:
I'm wondering whether we require all these methods individually. It would be beneficial to have an API for the front end that includes singular methods. These would be invoked according to user actions. We could compose different functions (from other files) here, but it would be ideal if the API/router level could mirror user actions.

Reply:
If it's as I've described, please correct me.

Reply:
They are required separately because the fetching from devhub happens in the front-end currently. And so the front-end needs to know whether it needs to fetch the full webhapp or only the UI or only the happ or nothing at all from a devhub host via remote calls in order to be able to install the app.